client/app/controllers/subscribe.js

```diff
@@ -0,0 +1,92 @@
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+import fetch from 'fetch';
+import ENV from 'labs-zap-search/config/environment';
+
+const tlds = ['.com', '.gov', '.edu', '.net', '.org'];
+
+export default class SubscribeController extends Controller {
+  lastEmailChecked = '';
+
+  emailAlreadyExists = false;
+
+  emailNeedsConfirmation = false;
+
+  startContinuouslyChecking = false;
+
+  emailSent = false;
+
+
+  @action
+  async checkExistingEmail(event) {
+    const email = event.target.value;
+    if (email === this.lastEmailChecked) { return; }
+    this.set('lastEmailChecked', email);
+    this.set('startContinuouslyChecking', true);
+
+    try {
+      const response = await fetch(`${ENV.host}/subscribers/email/${email}`);
+      const userData = await response.json();
+
+      if (userData.error) {
+        this.set('emailAlreadyExists', false);
+        this.set('emailNeedsConfirmation', false);
+        this.set('emailSent', false);
+        return;
+      }
+
+      if (userData.message === 'User already subscribed.') {
+        this.set('emailAlreadyExists', true);
+        this.set('emailNeedsConfirmation', false);
+      } else if (userData.message === 'User is subscribed but must confirm.') {
+        this.set('emailNeedsConfirmation', true);
+        this.set('emailAlreadyExists', false);
+      }
+      return;
+    } catch (error) {
+      // We will receive an error if:
+      // a) the user does not exist in Sendgrid, or
+      // b) their confirmed field is null.
+      // Either way, we don't need to log to console
+      this.set('emailAlreadyExists', false);
+      this.set('emailNeedsConfirmation', false);
+      this.set('emailSent', false);
+    }
+  }
+
+  @action
+  continuouslyCheckEmail(event) {
+    if ((this.startContinuouslyChecking) || (tlds.includes(event.target.value.slice(-4)))) { this.checkExistingEmail(event); }
```

Comment:
From testing this, it looks like it hits the API on every key press as soon as the user starts typing. This would be a good use case for implementing debouncing but that's maybe overkill. Instead, how about you validate the given value against the email regex and only call the API if it passes? That way we aren't spamming the API will calls that we know aren't going to find anything.

Reply:
The API isn't being called on every key press as soon as the user starts typing.  API calls begin when the end of the email input is one of the TLDs, or when the user removes focus from the email input.  Only after this point will API calls occur after every keystroke.  Because most users will enter their email address and then not change the field, it will minimize the number of queries, while allowing for immediate feedback for those that do change the field (which is useful for testing/development).  Replacing the TLDs with the email regex and logic used in `subscription-form.js` would allow queries to begin after any 2 letters after the period, meaning almost all users would make two queries when they first enter their address.

Let me know if you still would like me to switch things up.

Reply:
I'd still like to switch it out. I just think sticking with that regex is going to serve us better in the long run.